### PR TITLE
[codex] Fix mlx-community checkpoint loading

### LINKED ISSUE #12: Loader does not honor tie_word_embeddings (lm_head.weight missing for all mlx-community variants)
## Summary

All mlx-community Qwen3-ASR checkpoints store `lm_head.weight` as a **tied
copy of `model.embed_tokens.weight`** (`tie_word_embeddings: True` in
`config.json`). The library's loader expects `lm_head.weight` as a separately
stored tensor and fails with `ValueError: Missing 1 parameters: lm_head.weight`.

## Affected models

All `mlx-community/Qwen3-ASR-*` variants:

- `mlx-community/Qwen3-ASR-0.6B-bf16` / `8bit` / `6bit` / `5bit` / `4bit`
- `mlx-community/Qwen3-ASR-1.7B-bf16` / `8bit` / `6bit` / `5bit` / `4bit`

Verified by checking each repo's `config.json` (nested
`thinker_config.text_config.tie_word_embeddings` and
`audio_config.tie_word_embeddings` are both `True`) and
`model.safetensors.index.json` (no key containing `lm_head`).

## Versions

- `mlx-qwen3-asr` 0.3.3 and 0.3.4 (loader code is byte-identical between them)
- macOS / Apple Silicon

## Reproduction

\`\`\`python
from mlx_qwen3_asr import Session
sess = Session(\"mlx-community/Qwen3-ASR-1.7B-bf16\")
# ValueError: Missing 1 parameters: lm_head.weight.
\`\`\`

## Diagnosis

In \`mlx_qwen3_asr/load_models.py\`, \`_load_model_with_resolved_path\` does not
inspect \`tie_word_embeddings\` in the config or copy \`model.embed_tokens.weight\`
to \`lm_head.weight\` when the tensor is tied.

## Suggested fix

After remapping the weights, check the config and inject \`lm_head.weight\`
(plus \`.scales\` / \`.biases\` for quantized variants) from
\`model.embed_tokens\`:

\`\`\`python
def _has_tie(cfg_dict):
    if isinstance(cfg_dict, dict):
        for k, v in cfg_dict.items():
            if k == \"tie_word_embeddings\" and v: return True
            if isinstance(v, dict) and _has_tie(v): return True
    return False

if (_has_tie(raw_config)
        and \"lm_head.weight\" not in weights
        and \"model.embed_tokens.weight\" in weights):
    weights[\"lm_head.weight\"] = weights[\"model.embed_tokens.weight\"]
    if \"model.embed_tokens.scales\" in weights:
        weights[\"lm_head.scales\"] = weights[\"model.embed_tokens.scales\"]
    if \"model.embed_tokens.biases\" in weights:
        weights[\"lm_head.biases\"] = weights[\"model.embed_tokens.biases\"]
\`\`\`

For quantized checkpoints the \`class_predicate\` passed to \`nn.quantize\` must
also include \`lm_head\` in the quantized-paths set (which it will if
\`lm_head.scales\` is now present in the weights).

## Workaround for downstream users

Monkey-patch \`_load_model_with_resolved_path\` from your application code
until this is fixed upstream. (We have a working patch in our app.)

### LINKED ISSUE #13: Loader does not handle partially-quantized mlx-community checkpoints (audio_tower stays fp16)
## Summary

The mlx-community quantized Qwen3-ASR checkpoints quantize **only the
language decoder**; `audio_tower` (encoder) is left in full precision. The
library's loader detects quantization in the weights and calls
\`nn.quantize(model, bits=bits, group_size=group_size)\` on the **whole
model** without a \`class_predicate\`, which invents 297 \`.biases\` /
\`.scales\` slots in \`audio_tower\` that the safetensors never fill.

Result: \`ValueError: Missing 297 parameters: audio_tower.conv_out.biases, ...\`

## Affected models

All 8 quantized \`mlx-community/Qwen3-ASR-*\` variants:

- \`Qwen3-ASR-0.6B-4bit\` / \`5bit\` / \`6bit\` / \`8bit\`
- \`Qwen3-ASR-1.7B-4bit\` / \`5bit\` / \`6bit\` / \`8bit\`

bf16 variants are unaffected by this specific bug (but see #12 for the
tied-embeddings issue that also blocks them).

## Versions

- \`mlx-qwen3-asr\` 0.3.3 and 0.3.4 (loader unchanged between them)
- macOS / Apple Silicon

## Reproduction

\`\`\`python
from mlx_qwen3_asr import Session
sess = Session(\"mlx-community/Qwen3-ASR-1.7B-8bit\")
# ValueError: Missing 297 parameters: audio_tower.conv_out.biases, ...
\`\`\`

## Diagnosis

Inspection of \`model.safetensors.index.json\` for
\`Qwen3-ASR-1.7B-8bit\`:

| Submodule | \`.weight\` keys | \`.scales\` keys | \`.biases\` keys |
|---|---|---|---|
| \`audio_tower\` | 199 | 0 | 0 |
| \`model\` (decoder) | 310 | 197 | 197 |

So \`audio_tower\` is full-precision and only \`model.*\` is quantized.

In \`mlx_qwen3_asr/load_models.py\`:

\`\`\`python
quantized = _is_quantized_weights(weights)  # True (model.* has .scales)
if quantized:
    nn.quantize(model, bits=bits, group_size=group_size)  # quantizes EVERYTHING
\`\`\`

Without a \`class_predicate\`, \`nn.quantize\` converts every \`nn.Linear\` /
\`nn.Embedding\` in the model — including those under \`audio_tower\` — into
quantized layers expecting \`.scales\` / \`.biases\` that don't exist in the
safetensors.

## Suggested fix

Build the \`class_predicate\` from the weight keys themselves and pass it
to \`nn.quantize\`:

\`\`\`python
quantized_paths = {
    k.rsplit(\".scales\", 1)[0]
    for k in weights
    if k.endswith(\".scales\")
}
def _predicate(path, _module):
    return path in quantized_paths

if quantized:
    nn.quantize(model, bits=bits, group_size=group_size,
                class_predicate=_predicate)
\`\`\`

This way only the layer paths that actually have stored \`.scales\` get
converted to quantized layers; everything else stays in full precision and
the safetensors load cleanly.

## Workaround for downstream users

Monkey-patch \`_load_model_with_resolved_path\` from your application code
until this is fixed upstream. (We have a working patch in our app that
addresses both this issue and #12 together.)

### PR DESCRIPTION
**Pull request by moona3k**
## Summary

Fixes loading for `mlx-community` Qwen3-ASR checkpoints.

These checkpoints use tied `lm_head` and `model.embed_tokens` weights, and the quantized variants only quantize the decoder while keeping the audio encoder full precision. The old loader expected a separate `lm_head.weight` and quantized the whole model, so affected models failed before transcription started.

## What changed

- Fill missing tied `lm_head` tensors from `model.embed_tokens` when the config uses tied embeddings.
- Quantize only module paths that have saved `.scales` tensors in the checkpoint.
- Add regression tests for tied and partially quantized checkpoint layouts.

## Validation

- `uv run pytest tests/test_load_models.py -q`
- `uv run python scripts/quality_gate.py --mode fast`
- Real loader smokes: `mlx-community/Qwen3-ASR-0.6B-bf16`, `4bit`, and `8bit`

Fixes #12.
Fixes #13.

### CODE REVIEW

File: mlx_qwen3_asr/load_models.py [hunk 1]
```diff
@@ -160,6 +161,55 @@ def _cast_tree_dtype(tree: dict, dtype: mx.Dtype) -> dict:
...
+
+def _materialize_tied_lm_head_weights(
+    weights: dict[str, mx.array],
+    config: Qwen3ASRConfig,
+) -> dict[str, mx.array]:
+    """Fill missing LM-head tensors for checkpoints with tied embeddings."""
+    if not _model_uses_tied_lm_head(config):
+        return weights
+    if "lm_head.weight" in weights or "model.embed_tokens.weight" not in weights:
+        return weights
```
[the remark this answers is not in the document]
Comment (moona3k) [reply mlx_qwen3_asr/load_models.py]: Addressed in `02cd227`. The loader now handles the edge case where `lm_head.weight` is present but tied quantization aux tensors are missing: it fills missing `lm_head.scales` / `lm_head.biases` from `model.embed_tokens` when the existing LM-head weight shape matches the embedding weight shape. I also added regression coverage for that path and for the shape-mismatched no-op case.

File: mlx_qwen3_asr/load_models.py [hunk 2]
```diff
@@ -160,6 +161,55 @@ def _cast_tree_dtype(tree: dict, dtype: mx.Dtype) -> dict:
...
+        src = f"model.embed_tokens{suffix}"
+        dst = f"lm_head{suffix}"
+        if src in weights and dst not in weights:
+            weights[dst] = weights[src]
+    return weights
+
+
+def _quantized_module_paths(weights: dict[str, mx.array]) -> set[str]:
+    """Return module paths that have saved MLX quantization parameters."""
+    return {key.removesuffix(".scales") for key in weights if key.endswith(".scales")}
```
[the remark this answers is not in the document]
Comment (moona3k) [reply mlx_qwen3_asr/load_models.py:203]: Leaving this unchanged intentionally. MLX quantized checkpoints use `.scales` / `.biases`, and this loader already detects quantization through `_is_quantized_weights()` by looking for `.scales`. Supporting unknown quantization metadata names would be speculative and would risk quantizing modules without matching `load_weights()` tensors. The PR keeps the predicate tied to the concrete MLX checkpoint convention we can validate.

**Comment by moona3k** [comment]
Addressed the review follow-ups in `02cd227`.

- Hardened tied `lm_head` auxiliary tensor materialization for the explicit-weight edge case.
- Added regression tests for missing aux tensors and shape-mismatched explicit projections.
- Added small helper docstrings for the docstring coverage warning.
- Re-ran `uv run pytest tests/test_load_models.py -q` and `uv run python scripts/quality_gate.py --mode fast`.
